pallets/manta-pay/src/lib.rs

```diff
@@ -480,19 +465,140 @@ pub mod pallet {
 	where
 		T: Config,
 	{
-		/// Returns the ledger implementation for this pallet.
+		/// Maximum Number of Updates per Shard
+		const PULL_MAX_PER_SHARD_UPDATE_SIZE: usize = 128;
+
+		/// Maximum Size of Sender Data Update
+		const PULL_MAX_SENDER_UPDATE_SIZE: usize = 1024;
+
+		/// Pulls receiver data from the ledger starting at the `receiver_index`.
 		#[inline]
-		fn ledger() -> Ledger<T> {
-			Ledger(PhantomData)
+		fn pull_receivers(receiver_index: &mut [usize; 256]) -> (bool, ReceiverChunk) {
+			let mut more_receivers = false;
+			let mut receivers = Vec::new();
+			for (i, index) in receiver_index.iter_mut().enumerate() {
+				more_receivers |= Self::pull_receivers_for_shard(i as u8, index, &mut receivers);
```

Comment:
while this is compact and *clever* logic, I was wondering where `receivers` actually gets filled.
I'd prefer a simple multiline solution over this compact one

Can `pull_receivers_for_shard` not return an `Option<ReceiverChunk>`?

Reply:
`receivers` gets filled in `pull_receivers_for_shard`. This is a very common trick even in `C/C++`: pass a mutable reference used as return while returning a boolean type as a flag.

Reply:
yeah, in C++ out-references are common to the point of becoming an anti-pattern people are steered away from. e.g. https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f21-to-return-multiple-out-values-prefer-returning-a-struct-or-tuple

the idea is that a return value is simply *the right way* to express something that's clearly the output of a function as opposed to just a side-effect, where an out-reference hides that. This must also be possible for rust. Let me think about how I would rewrite this.

Reply:
Also, isnt it just *wrong* to use `|=` here? If any of the calls returns `true`, `more_receivers` will always be `true` at the end of the call, ever if a subsequent loop iteration returns `false`

Reply:
> Also, isnt it just _wrong_ to use `|=` here? If any of the calls returns `true`, `more_receivers` will always be `true` at the end of the call, even if a subsequent loop iteration returns `false`

That is the intended behavior. To indicate the client/wallet that it needs to keep pulling since there are more diff to pull.

Reply:
I didn't write this code, but the code is quite intuitive to me.

Reply:
the main gripe i have is that the fn returns the unimportant bool flag instead of its actual `receivers` vector.
i tried out a couple of tuple-based returns, and am unhappy with them too, so I'm ok keeping the current implementation

Reply:
> the main gripe i have is that the fn returns the unimportant bool flag instead of its actual `receivers` vector. i tried out a couple of tuple-based returns, and am unhappy with them too, so I'm ok keeping the current implementation

Why the bool flag is unimportant? It is quite important in my view.

Reply:
the primary purpose of the fn is to return the `receivers`, right? the flag is of seconary importance to trigger a second round of reading. With the chosen signature, the bool flag is the return value, signifying higher importance than some out param